### PR TITLE
Calls an unrecallable shuttle once Gangs have overrun the station

### PR DESCRIPTION
**Pull request by Robustin**
If more than 75% of the players are dead or part of a gang then its time to put up or shut up.

You've still got several minutes to drop a dominator, which prevents the shuttle from leaving.

### CODE REVIEW

File: code/game/gamemodes/gang/gang.dm [hunk 1]
```diff
@@ -247,6 +248,20 @@ var/list/gang_colors_pool = list("red","orange","yellow","green","blue","purple"
 		gang_bosses += G.bosses
 	return gang_bosses
 
+/datum/game_mode/proc/shuttle_check()
+	if(forced_shuttle)
+		return
+	var/normies = 0 
+	for(var/mob/living/M in GLOB.player_list)
+		if(isliving(M) && !M.stat==DEAD && (!M.mind in get_all_gangsters())
```
Comment (Incoming5643) [inline code/game/gamemodes/gang/gang.dm]: You'd be running get_all_gangsters a lot less if you just assigned its contents to a list outside the for loop.

Comment (Robustin) [inline code/game/gamemodes/gang/gang.dm]: I'll add a global list, perhaps. 

Keep in mind it won't be accessed THAT often though, this check only happens when the Gang collects points (probably the best time for a Gang to make determinations about saving for a dominator).

Comment (RemieRichards) [inline code/game/gamemodes/gang/gang.dm]: @Robustin the point is you're running ```get_all_gangsters()``` for every single player in the player list (which is all of them)

There's literally no reason for this, just where you have ```var/normies = 0``` outside the loop, you can define: ```var/list/cool_dudes = get_all_gangsters()``` and then you can change this to: ```if( ... && (!M.mind in cool_dudes))```

No functionality changes, Much more efficient code.

File: code/game/gamemodes/gang/gang.dm [hunk 2]
```diff
@@ -247,6 +248,20 @@ var/list/gang_colors_pool = list("red","orange","yellow","green","blue","purple"
 		gang_bosses += G.bosses
 	return gang_bosses
 
+/datum/game_mode/proc/shuttle_check()
+	if(forced_shuttle)
+		return
+	var/normies = 0 
+	for(var/mob/living/M in GLOB.player_list)
+		if(isliving(M) && !M.stat==DEAD && (!M.mind in get_all_gangsters()))
```
Comment (RemieRichards) [inline code/game/gamemodes/gang/gang.dm]: and also, the ```isliving(M)``` isn't necessary, as the fact you've typecasted the loop as ```for(var/mob/living/M in GLOB.player_list)``` means that you'll only get ```/mob/living```s or subtypes of ```/mob/living```.

Also ```M``` should be ```L``` as this is a ```/living``` not a ```/mob```, just consistency.

**Comment by KorPhaeron** [comment]
This conflicts pretty heavily with the no sec/armory thing, do you no longer plan on doing that?

**Comment by optimumtact** [comment]
gang was removed

**Comment by Robustin** [comment]
Kor, I intended for this mechanic to work under the current system. If we moved ahead with no sec then I'd have to tweak the numbers but the message is still critical: Killing everyone and THEN dropping the dominator should never be a preferred strategy. If the round is heading in that direction this mechanic is a safety valve to force the gangs to make their dominator play before they've wiped out all resistance.

With sec gone I'd put more emphasis on checking if a single gang is dominant.

**Comment by KorPhaeron** [comment]
@Robustin 

It's technically possible to have killed no one and still have this trigger. If you really want a check like this it should be based on an individual gangs size, not short circuiting the round because there are three large gangs.

**Comment by KorPhaeron** [comment]
@lzimann 

Do you still have the code for the armory/sec removal during gang I probably shouldn't be arguing on theoretical changes

**Comment by KorPhaeron** [comment]
I do think a change as drastic as that is the only way to salvage gang though, anything else will just turn into sec buying insane amounts of implants with the broken cargo economy and implanting the entire crew.

**Comment by PKPenguin321** [comment]
This is smart though. It forces the finale scenario so either combat has to happen after the dominator is placed rather than before, or everybody's dead and the round isn't able to drag on for no reason. I like it a lot

File: code/game/gamemodes/gang/gang.dm (hunk 2 above)
Comment (Cyberboss) [inline code/game/gamemodes/gang/gang.dm]: There's more wrong with this, !M.stat, !M.mind, and calling get_all_gangsters every iteration

Change the whole bit to:
```
var/list/all_gangsters = get_all_gangsters()
for(var/mob/living/L in GLOB.player_list)
    if(M.stat != DEAD && !(L.mind in all_gangsters))
```

**Comment by lzimann** [comment]
I don't, but it is something that takes 5 minutes to do. I won't do it though.

**Comment by Robustin** [comment]
Kor if there's 3 large gangs, then there's 3 gangs who can all drop dominators and make the shuttle a non-issue.

If there's 3 large gangs I can't think of a better way to shift to the end-game  than forcing each gang to drop their Dom and descend into a crazy 3-way fight. That's far better than the alternative of "lets hunker down and enter a war of attrition for 45 minutes because nobody wants to be the first to drop a dom and become a focus for the other 2 gangs". 

@lzimann 

Sure, but if you could link the code or even describe the method, it would be nice.

**Comment by Iamgoofball** [comment]
it's easy

gamemode code occurs before the game assigns jobs, so all you have to do is:
1. loop through areas for the Armory/Security areas, delete all guns with a quick istype for the gun path
2. remove the security job datums from the job subsystem

**Comment by Lumipharon** [comment]
Good change.
If most of the crew is dead or gang, there is no room for further round progression outside:
A: dropping a dong
B: Killing every last non gang buddy for 100% guaranteed success

Which option is preferable for an more interesting round? Hell, even making it lower than 70% would be good.

**Comment by lzimann** [comment]
ping when this compiles I guess

**Comment by KorPhaeron** [comment]
This is not compatible with the security removal PR, will result in shuttles at 10 minutes in

**Comment by Robustin** [comment]
Then we tweak the numbers until it is.

There needs to be a safety valve against zombie rounds.  Without security nobody even wants to leave until they win and nobody wants to attempt to win until they've wiped out anyone who can stop them.

**Pull request closed without merging**